src/neo/Network/P2P/Payloads/Signer.cs

```diff
@@ -45,17 +47,23 @@ public class Signer : ISerializable
         /// </summary>
         public ECPoint[] AllowedGroups;
 
+        /// <summary>
+        /// The rules that the witness must meet. Only available when the <see cref="WitnessScope.WitnessRules"/> flag is set.
+        /// </summary>
+        public WitnessRule[] Rules;
+
         public int Size =>
             /*Account*/             UInt160.Length +
             /*Scopes*/              sizeof(WitnessScope) +
             /*AllowedContracts*/    (Scopes.HasFlag(WitnessScope.CustomContracts) ? AllowedContracts.GetVarSize() : 0) +
-            /*AllowedGroups*/       (Scopes.HasFlag(WitnessScope.CustomGroups) ? AllowedGroups.GetVarSize() : 0);
+            /*AllowedGroups*/       (Scopes.HasFlag(WitnessScope.CustomGroups) ? AllowedGroups.GetVarSize() : 0) +
```

Comment:
I think that we can remove `AllowedContracts` and `AllowedGroups` and convert it into rules during deserialization.

Reply:
We can't. If we do so, the hash of the transaction is changed.